#20153 #20152 #20155 #20210: Collection of small EDM optimizations

tt-aho opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/20153
https://github.com/tenstorrent/tt-metal/issues/20152
https://github.com/tenstorrent/tt-metal/issues/20155
https://github.com/tenstorrent/tt-metal/issues/20210

### Problem description
We're trying to shave off/stabilize perf numbers for ring on 1D fabric.
We want to set a higher ETH_TXQ_DATA_PACKET_ACCEPT_AHEAD which is expected to increase perf stability, reduce setting did_something setting in most branches except for when we are forwarding/sending in sender/receiver, and optimize trid calculations when trid count == buffer slots.

### What's changed
Set ETH_TXQ_DATA_PACKET_ACCEPT_AHEAD to 32 for edm fabric
Refactor did_something checks to only set to true when either sender or receiver is forwarding/writing data
Clean up trid handling for pow2/non-pow2 cases, and add handling for optimized case where num trids == num slots
Testing with increasing num trids to 8 to match number of buffer slots results in a large perf regression in the true ring ubench tests, so this is dropped.
Add reference results for 6u.

<img width="407" alt="image" src="https://github.com/user-attachments/assets/2857f02e-7da2-4bea-bcf1-5fe2edd21d7d" />

Overall net positive except for the non-fwd test, which seems to have been negatively affected by the did_something changes.


### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions?query=actor%3Att-aho
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] **(For models and ops writers)** Full [new models tests](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) CI passes (if applicable)
- [x] New/Existing tests provide coverage for changes

UBench: https://github.com/tenstorrent/tt-metal/actions/runs/14359939024
T3K Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/14359949336
TG Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/14359953802
T3K Unit: https://github.com/tenstorrent/tt-metal/actions/runs/14365587681

tt-aho committed:
#20153: Set ETH_TXQ_DATA_PACKET_ACCEPT_AHEAD to 32 for edm fabric

tt-aho committed:
#20152: Refactor did_something checks to only set to true when either sender or receiver is forwarding/writing data

tt-aho committed:
#20155: Clean up trid handling for pow2/non-pow2 cases, and add handling for optimized case where num trids == num slots

SeanNijjar commented on tt_metal/fabric/hw/inc/edm_fabric/1d_fabric_transaction_id_tracker.hpp:
```diff
@@ -35,24 +35,28 @@ struct WriteTransactionIdTracker {
+    static_assert(OFFSET + MAX_TRANSACTION_IDS - 1 <= NOC_MAX_TRANSACTION_ID, "Invalid transaction ID");
 
     WriteTransactionIdTracker() {
         for (size_t i = 0; i < NUM_CHANNELS; i++) {
```
can go away for pow2 or if they are same.

SeanNijjar commented on tt_metal/fabric/hw/inc/edm_fabric/1d_fabric_transaction_id_tracker.hpp:
```diff
@@ -35,24 +35,28 @@ struct WriteTransactionIdTracker {
     update_buffer_slot_to_next_trid_and_advance_trid_counter(tt::tt_fabric::BufferIndex buffer_index) {
-        if constexpr (BOTH_PARAMS_ARE_POW2) {
+        if constexpr (BOTH_PARAMS_ARE_EQUAL) {
+            uint8_t next_trid = OFFSET + buffer_index;
```
I think we can just omit this code altogether since the #trids == #slots, there isn't really a "next" trid.

tt-aho committed:
#20210: Update ubench to run on 6u and add reference golden file for 6u

SeanNijjar commented on tt_metal/fabric/hw/inc/edm_fabric/1d_fabric_transaction_id_tracker.hpp:
```diff
@@ -63,38 +69,33 @@ struct WriteTransactionIdTracker {
-        for (uint8_t i = 0; i < NUM_CHANNELS; ++i) {
-            tt::tt_fabric::BufferIndex buffer_index(i);
-            auto trid = this->get_buffer_slot_trid(buffer_index);
+        for (uint8_t i = 0; i < MAX_TRANSACTION_IDS; ++i) {
```
nice - yea the old code was a bit redundant for the non-pow2 case.

tt-aho committed:
#0: Fix race condition in 1D fabric teardown from device init
There was a race where we issue inline writes which will receive an ack back, but we don't barrier on it and EDM force resets the noc counters, so the sw/hw counters go out of sync
